scripts/core/editor3/components/spellchecker/Spellchecker.jsx

```diff
@@ -0,0 +1,229 @@
+/**
+ * @ngdoc React
+ * @module superdesk.core.editor3
+ * @name Spellchecker
+ * @param {Function} getEditorState
+ * @param {Function} setEditorState
+ * @param {Object} spellchecker
+ * @param {Object} editorRect
+ * @description check the text and shows the inline spellchecker context menu
+ */
+
+import React from 'react';
+import {EditorState, Modifier} from 'draft-js';
+
+export class Spellchecker extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.addWord = this.addWord.bind(this);
+        this.replaceWord = this.replaceWord.bind(this);
+
+        this.isVisible = this.isVisible.bind(this);
+    }
+
+    /**
+      * @ngdoc method
+      * @name Spellchecker#getDecorator
+      * @param {Function} getEditorState
+      * @param {Function} setEditorState
+      * @param {Object} spellchecker
+      * @return {Object} returns the decorator
+      * @description Decorator for spellchecker, will highlight the words with errors
+      */
+    static getDecorators(getEditorState, setEditorState, spellchecker) {
+        return [{
+            strategy: spellcheckStrategy(spellchecker.isCorrectWord),
+            component: Spellchecker.spellcheckSpan,
+            props: {
+                getEditorState: getEditorState,
+                setEditorState: setEditorState,
+                spellchecker: spellchecker,
+                showContextMenu: Spellchecker.showContextMenu
+            }
+        }];
+    }
+
+    /**
+      * @ngdoc method
+      * @name Spellchecker#spellcheckSpan
+      * @return {Object} return the component used to highlight errors
+      * @description Component to highlight errors
+      */
+    static spellcheckSpan(props) {
+        const style = styleMap.ERROR;
+        const onContextMenu = (e) => props.showContextMenu(e, props);
+
+        return <span style={style} onContextMenu={onContextMenu}>{props.children}</span>;
+    }
+
+    /**
+      * @ngdoc method
+      * @name Spellchecker#showContextMenu
+      * @param {Object} e on context menu event
+      * @param {Object} the properties of the component that generates the event
+      * @description Show the contextual menu, init the suggestions list
+      */
+    static showContextMenu(e, props) {
+        const left = e.clientX;
+        const top = e.clientY;
+        var element = props.children[0];
+
+        e.preventDefault();
+        props.spellchecker.suggest(element.props.text)
+        .then((suggestions) => {
+            var {editorState} = props.getEditorState();
+
+            props.setEditorState({
```

Comment:
This is the spellchecker's state, which is this component. The spellchecker should manage the spellchecker's state. Here, you are using a static method of the spellchecker, that takes a function passed as a property, which belongs to the editor, that changes the spellcheckers state. This is very complex.

I think that the creators of draft-js intended decorators to not have access to any state or props. But I might be wrong!

I'm new and I'm not sure what the best pattern is, but this still seems like it good be simplified. I would like to suggest the following method:

* Create an [`Entity`](https://facebook.github.io/draft-js/docs/api-reference-entity.html#create) of type `TYPO` wherever there is a spelling issue.
* Because `editorState` is a prop of Spellchecker, it will cause a re-render every time anything changes. You can check the current cursor position to see if it's on a `TYPO` entity and display the context menu in that situation. I have done the same in toolbar. There are entities of type `LINK` and when the cursor is over them, the popover is displayed.

I think doing the above method will isolate all of the components and they will not depend on each other anymore. Their only common denominator will be `editorState`. It will also rid you of all the callbacks.